File: gcj/gcj/188266/Marquis/168107/0/extracted/multibase.py
#!/usr/bin/python

def get_digits(n, base):
    while n:
        yield n % base
        n = n / base

def is_happy(n, base):
    seen = set([n])
    
    while 1:
        new = 0
        for i in get_digits(n, base):
            new += i*i
            
        # print new
        n = new
        if n in seen: return False
        if n == 1: return True
        seen.add(n)

t = int(raw_input())

for i in range(0, t):
    bases = map(int, raw_input().split())
    n = 2
    done = False
    while not done:
        yes_in_all = True
        for base in bases:
            if not is_happy(n, base):
                yes_in_all = False
                break
        if yes_in_all:
            print "Case #"+str(i+1)+":",n
            done = True
        n += 1
         

File: gcj/gcj/188266/mcphee/168107/0/extracted/MultiBase.rb
#! /usr/bin/ruby

def happy(x, base, seen)
  # puts "Checking #{x} for base #{base} having seen #{seen}\n"
  if seen.member?(x)
    return false
  end
  seen.push(x)
  str = x.to_s(base)
  ss = 0
  str.each_byte do |c|
    d = c - '0'[0]
    ss += d*d
  end
  if ss == 1
    return true
  end
  return happy(ss, base, seen)
end

def all_happy(x, bases)
  # puts "Checking #{x} for bases #{bases}\n"
  return bases.all? do |base|
    happy(x, base, [])
  end
end

filename = ARGV[0]
file = File.new(filename)
lines = file.readlines

lines.shift

lines.each_with_index do |l, c|
  bases = l.split(" ").map { |s| s.to_i }
  x = 2
  while (not all_happy(x, bases)) do
    # puts "x = #{x}\n"
    x = x + 1
  end
  puts "Case \##{c+1}: #{x}\n"
end

